### PR TITLE
zipLatest and zipLatestWith operators (refs #25392)

### PR DESCRIPTION
**Pull request by jchapuis**
Adds `ZipLatestWith` GraphStage and corresponding `zipLatest` and `zipLatestWith` operators. 

Similar to `zip` and `zipWith`, these operators combine elements from multiple sources. `zipLatest` zips them into a stream of tuples, `zipLatestWith` applies a combiner function on the tuple. 

In contrast to `zip` and `zipWith` however, `zipLatest` and `zipLatestWith` emit each time an element becomes available on either of the inputs, picking always the latest element of each. Note that no element is emitted until at least one element of each becomes available.

These operators have similar semantics to ReactiveX's `combineLatest` (http://reactivex.io/documentation/operators/combinelatest.html). It allows zipping streams "in time", which is for instance useful for aggregating a unified status from streams with varying rates.

### CODE REVIEW

**Comment by jchapuis** [comment]
Looks like build fails due to some binary compatibility check and the fact that the PR introduces some new operators in `akka.stream.scaladsl.FlowOpsMat` for instance ? What should I do? Follow the suggestion and filter them out?

**Comment by jchapuis** [comment]
Binary compatibility errors are:

```[error]  * method zipLatestMat(akka.stream.Graph,scala.Function2)akka.stream.scaladsl.FlowOpsMat in trait akka.stream.scaladsl.FlowOpsMat is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOpsMat.zipLatestMat")
[error]  * method zipLatestWithMat(akka.stream.Graph,scala.Function2,scala.Function2)akka.stream.scaladsl.FlowOpsMat in trait akka.stream.scaladsl.FlowOpsMat is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOpsMat.zipLatestWithMat")
[error]  * method zipLatestGraph(akka.stream.Graph)akka.stream.Graph in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestGraph")
[error]  * method zipLatest(akka.stream.Graph)akka.stream.scaladsl.FlowOps in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatest")
[error]  * method zipLatestWithGraph(akka.stream.Graph,scala.Function2)akka.stream.Graph in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestWithGraph")
[error]  * method zipLatestWith(akka.stream.Graph,scala.Function2)akka.stream.scaladsl.FlowOps in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestWith")```

**Comment by raboof** [comment]
Yes: you'll notice the comment above `FlowOpsMat` and `FlowOps` mention that these are 'internal API', and binary compatibility only needs to be maintained for callers of those classes, not for implementors, so it looks like you can safely filter those out.

**Comment by jchapuis** [comment]
ready for review.

**Review by johanandren — COMMENTED** [review]
Looks promising! Added some questions.

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipLatest.md
Comment (jchapuis) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipLatest.md]: yes I would tend to agree, however I followed the same convention as for 'zip.md' and 'zipWith.md' where the same paragraph is repeated between summary and description. So I guess either we would have to eliminate the description section (but it would probably look strange in the documentation) or come up with a more detailed explanation in the description?

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: I have tried but unfortunately (but also rather logically), OptionVal has a reference type constraint '>: Null', which means that this would propagate to 'ZipLatest'. Do you see an alternative or is there another version that supports value types? I find meta-programming limits the ugliness somewhat 😉

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: so it turns out that 'waitingForTuple' indicates that a pull has been done downstream, and 'hasAllValues' simply indicates that we have all values necessary to form up a tuple. 'hasAllValues' stays 'true' after emission of the first tuple. Therefore, even without pulling upstream here, a pull downstream will trigger emission of the next tuple (since 'onPull' triggers 'pullAllIfNeeded')

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: Actually this check is to avoid re-emitting another, identical tuple directly upon pull. The corresponding test is: `does not emit the same pair upon two pulls`

The scenario you describe should be covered, see the corresponding `emit even if pair is the same` test. 

Let me know if you want me to add more to these tests.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (johanandren) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: In practice all elements going through the stream will be boxed anyway, I thought there were stages where we use that knowledge and do some casting to AnyRef internally, but now I can't find such a stage, so maybe those has been eliminated.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (johanandren) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: I think it'd need to be the same instance of pair, not an equal pair then?

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipLatest.md
Comment (jchapuis) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipLatest.md]: yes, good catch

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestWithSpec.scala
Comment (jchapuis) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestWithSpec.scala]: cannot agree more, changing

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: yes, I had noticed that too but did not think of adding it. I agree, for the sake of consistency it's probably better to have the same behavior, adding.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestWithSpec.scala
Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestWithSpec.scala]: Great, thanks!

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: yes indeed (although I will have to reference the outer type), changing this

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: ok, changing

Comment (johanandren) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: I think you'd have to do some somewhat uggly `asInstanceOf[AnyRef]` on the elements and use `OptionVal[AnyRef]` but on second thought, maybe it isn't worth it. Let's keep it as is for now.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: yes, adding

**Comment by jchapuis** [comment]
Ok, I did my best to improve it based on your good comments, ready for another pass. I can squash the commits once we're done with the review.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (johanandren) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: That's still equality check though? For instance equality it would have to be `lastOutput.ne(output)`.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: So to cover the instance equality this would be `true` `true` on source one and 1 1 1 on source 2. And you would see three tuples emitted, because there was three separate elements coming in.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (johanandren) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: And I wonder if we _really_ need to keep track of the previous tuple at all, if we don't do said equality based deduplication.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala
Comment (jchapuis) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: this is what the equality comparison is for: we don't want to re-emit the same output when pulling again on the source. But we don't dedup at all when new values are pushed from upstream.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala
Comment (jchapuis) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: likewise here, added a separate test for reference types

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala
Comment (jchapuis) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: ok I will change this test to cover this

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala
Comment (jchapuis) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: this is what the equality comparison is for: we don't want to re-emit the same output when pulling again on the source. But we don't dedup when new values are pushed from upstream.

Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: The question here is what the "same" is. I argue it is the same instance of tuple, not a tuple that is equal to a previously created tuple. Since it shouldn't be about equality I think we can get away with tracking the fact that a value came in and we haven't emitted anything yet. (Potentially much cheeper than keeping the last emitted element in memory)

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: see my other inline comment on the corresponding test: this check is for the case when we pull again on the source and nothing happened upstream: we don't want to re-emit again

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala
Comment (jchapuis) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: yes you are right. It would also be clearer. I'll see about modifying this, hopefully this afternoon

Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala]: Great, thanks!

**Comment by jchapuis** [comment]
ok I simplified the implementation and added some coverage, thanks for the great suggestions, let me know what you think

**Comment by jchapuis** [comment]
Hi @johanandren, did you have a chance to have a look at the changes? I think it's more streamlined now

**Comment by jchapuis** [comment]
not sure what just happened there? I didn't do any commits on the branch. A build was somehow trigged, which fails with `[09-18 09:49:22.713] [error] java.lang.RuntimeException: Error writing content for page cluster-usage.html: property [snip.code.base_dir] is not defined`

**Comment by ktoso** [comment]
Hey @jchapuis, don't worry about this failure. We had accidentally merged something that failed master so all PRs would fail as well. This was fixed right away though; nothing to do with your PR, we'll get to reviewing it as soon as we can, thanks / sorry! :)

**Comment by johanandren** [comment]
OK TO TEST

**Comment by jchapuis** [comment]
Failure seems related to doc reference ```(akka-docs / Compile / paradoxMarkdownToHtml) Error writing content for page stream/operators/Source-or-Flow/zipLatest.html: Unknown snippet [/localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/src/main/paradox/stream/operators/Source-or-Flow/$akka$/akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (No such file or directory)] referenced from [stream/operators/Source-or-Flow/zipLatest.html]```

**Comment by johanandren** [comment]
I think the way we reference sources to include in snippets has changed while this PR has been in flight, should work just by removing `$akka` in the snippet path and having a relative path from the root of the akka project.

**Comment by jchapuis** [comment]
Ok, that seemed to have done the trick.

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by jchapuis** [comment]
Travis CI fails with ```Downloading sbt launcher for 1.2.3:
  From  http://repo.scala-sbt.org/scalasbt/maven-releases/org/scala-sbt/sbt-launch/1.2.3/sbt-launch.jar
    To  /home/travis/.sbt/launchers/1.2.3/sbt-launch.jar
Download failed. Obtain the jar manually and place it at /home/travis/.sbt/launchers/1.2.3/sbt-launch.jar```

**Comment by johanandren** [comment]
Thanks (in general, no need to squash though, we can do that through the UI when merging)

**Comment by jchapuis** [comment]
hi there, do you need anything still from me to move this forward? let me know

**Comment by johanandren** [comment]
Still waiting for another review and LGTM from the core team, sorry for the delay, busy days.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (patriknw) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: How does this work? Isn't there a risk that we will later `pull` from a finished inlet? Or is that guarded against via `hasBeenPulled`? Otherwise we might want to use `tryPull`.

**Review by patriknw — COMMENTED** [review]
looking good, one question about onPull after onUpstreamFinish

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (jchapuis) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: thanks for this comment, I have had a closer look at how termination is handled and realized that one case was not covered: the case when a first tuple is ready for emission and we have an upstream completion right after: in the zip operator, semantics are to still emit the first tuple. 

So I've changed the semantics slightly to match this behavior and added the corresponding coverage.

Regarding your question, I don't think this case can arise since whenever we have stale input values (which is also the case initially), we complete the stage immediately, and in the other case we will will complete upon the following pull.

But to be on the safe side and since the semantics are in essence closer to what it's supposed to do, I have changed the call to 'tryPull'

**Comment by johanandren** [comment]
Existing zip should be fixed for consistency, but I don't think anyone noticed it before which could be an indication on how urgent the fix is ;)

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template
Comment (patriknw) [reply akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template]: thanks

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
> Existing zip should be fixed for consistency

I don't think existing zip has the same problem, because it will not pull at all after `willShutDown = true`

```
    private def pushAll(): Unit = {
      push(out, zipper(grab(in0), grab(in1)))
      if (willShutDown) completeStage()
      else {
        pull(in0)
        pull(in1)
      }
    }
```

**Comment by patriknw** [comment]
ah it was about

> overloads only support up to 19 parameters 

yes, could you fix that while you are at it @jchapuis ?

**Comment by jchapuis** [comment]
sure, will create a separate PR

**Review by johanandren — APPROVED** [review]
LGTM, thanks for the hard work :)

**Comment by johanandren** [comment]
Fixes #25392

**Pull request merged**
